etl/command.py

```diff
@@ -357,16 +376,53 @@ def run_dag(
             f"--- Would run {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
         )
         return enumerate_steps(steps)
-    elif workers == 1:
-        print(
-            f"--- Running {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
-        )
-        return exec_steps(steps, strict=strict)
-    else:
-        print(
-            f"--- Running {len(steps)} steps with {workers} processes ({config.GRAPHER_INSERT_WORKERS} threads each):"
-        )
-        return exec_steps_parallel(steps, workers, dag=dag, strict=strict)
+
+    if engine == "etl":
+        if workers == 1:
+            print(
+                f"--- Running {len(steps)} steps{_create_expected_time_message(total_expected_time_seconds, prepend_message=' (at least ')}:"
+            )
+            return exec_steps(steps, strict=strict)
+        else:
+            print(
+                f"--- Running {len(steps)} steps with {workers} processes ({config.GRAPHER_INSERT_WORKERS} threads each):"
+            )
+            return exec_steps_parallel(steps, workers, dag=dag, strict=strict)
+
+    elif engine == "prefect":
+        if workers == 1:
+            # task_runner = task_runners.SequentialTaskRunner()
+            task_runner = prefect.task_runners.ConcurrentTaskRunner()
+        else:
+            task_runner = DaskTaskRunner(cluster_kwargs={"n_workers": workers, "threads_per_worker": 1})
+
+        def on_failure_hook(task: prefect.Task, task_run: TaskRun, state: prefect.State):
+            """Cancel all tasks if one fails."""
+            with get_dask_client() as client:
+                print("Failure detected. Shutting down Dask client and cancelling all tasks.")
+                # force kill Dask client
+                client.shutdown()
+
+        @prefect.flow(log_prints=True, task_runner=task_runner)
+        def run_etl(steps):
+            task_futures: Dict[str, PrefectFuture] = {}
+
+            for step in steps:
+                # task = prefect.task(name=str(step), on_failure=[on_failure_hook])
+                task = prefect.task(name=str(step))
+
+                # include dependencies that are in the list of steps
+                wait_for = [task_futures[str(dep)] for dep in step.dependencies if str(dep) in task_futures]
+
+                def run_step(grapher_insert_workers=config.GRAPHER_INSERT_WORKERS):
+                    # We dynamically update `GRAPHER_INSERT_WORKERS` based on number of processes, the change in config
+                    # is not propagated when Dask creates a new process. We have to set it inside each task.
+                    config.GRAPHER_INSERT_WORKERS = grapher_insert_workers
+                    return step.run()
+
+                task_futures[str(step)] = task(run_step).submit(wait_for=wait_for)  # type: ignore
```

Comment:
You should have a separate run method for each engine, rather than a massive block like this.